### PR TITLE
Fix memory leak when compressing HIBA grant extensions and refactor decompression into extensions.c

### PR DESCRIPTION
**Pull request by blunderer**

### CODE REVIEW

File: extensions.c [hunk 1]
```diff
@@ -263,14 +263,18 @@ hibaext_maybe_deflate(struct sshbuf *in, int compress, struct sshbuf *out) {
...
 		ret = HIBA_OK;
 		debug3("hibaext_maybe_deflate: compression out %zu.", sshbuf_len(temp));
 	}
 err:
 #endif  /* WITH_EXTENSION_COMPRESSION */
 
 	if (ret == HIBA_OK) {
-		return sshbuf_putb(out, temp);
+		ret = sshbuf_putb(out, temp);
+		if (temp != in)
```
Comment (hallgrimurhg) [inline extensions.c:275]: Shouldn't it also be free'd for ret != HIBA_OK?

Comment (blunderer) [reply extensions.c]: Done

File: extensions.c [hunk 2]
```diff
@@ -263,14 +263,18 @@ hibaext_maybe_deflate(struct sshbuf *in, int compress, struct sshbuf *out) {
...
+		deflateEnd(&stream);
 		ret = HIBA_OK;
 		debug3("hibaext_maybe_deflate: compression out %zu.", sshbuf_len(temp));
 	}
 err:
 #endif  /* WITH_EXTENSION_COMPRESSION */
 
 	if (ret == HIBA_OK) {
-		return sshbuf_putb(out, temp);
+		ret = sshbuf_putb(out, temp);
```
Comment (hallgrimurhg) [inline extensions.c]: nit: sshbuf_putb returns an SSH error code, whereas hibaext_maybe_deflate returns a HIBA error code. The success code happens to be the same (0), but the error codes do not match (e.g. sshbuf_putb can return SSH_ERR_ALLOC_FAIL = -2, which corresponds to HIBA_BAD_VERSION).

File: extensions.c [hunk 3]
```diff
@@ -135,6 +225,104 @@ hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
 	return ret;
 }
 
+static int
+hibaext_decode_one(struct hibaext ***exts, int nexts, const unsigned char *data, size_t len) {
+	int ret;
+	struct sshbuf *blob = sshbuf_new();
+	sshbuf_put(blob, data, len);
```
Comment (hallgrimurhg) [inline extensions.c:232]: nit: check sshbuf_put return value

File: extensions.c [hunk 4]
```diff
@@ -135,6 +225,104 @@ hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
...
+			if (data[i] == ',') {
+				extension_len = i - pos;
+			} else if (i == len - 1) {
+				extension_len = len - pos;
+			}
+
+			if (extension_len > 0) {
+				debug3("hibaext_decode_all: found extension [%d, %zu]: %.*s", pos, i, extension_len, data+pos);
+				if ((ret = hibaext_decode_one(exts, nexts, data+pos, extension_len)) < 0) {
+					debug3("hibaext_decode_all: hibaext_decode_one returned %d: %s", ret, ssh_err(ret));
```
Comment (hallgrimurhg) [inline extensions.c:277]: Are decoding errors intentionally not treated as hard errors when calling hibaext_decode_all? The ret = nexts at the end overrides any error set here.

File: extensions.c [hunk 5]
```diff
@@ -135,6 +225,104 @@ hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
...
+		}
+	} else if (magic == HIBA_MAGIC) {
+		/* One extension using raw format. */
+		debug3("hibaext_decode_all: found 1 raw extension");
+		if ((ret = hibaext_decode_one(exts, nexts, data, len)) < 0) {
+			debug3("hibaext_decode_all: hibaext_decode_one returned %d: %s", ret, ssh_err(ret));
+		}
+		nexts++;
+	} else if (magic == HIBA_MULTI_EXTS) {
+		/* Consume magic MULT header. */
```
Comment (hallgrimurhg) [inline extensions.c:293]: nit: MULT => MULTI

File: extensions.c [hunk 6]
```diff
@@ -40,6 +42,94 @@ struct hibaext {
...
+		}
+	}
+	if (temp != in) {
+		sshbuf_free(temp);
+	}
+	return ret;
+}
+
 int
 hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
```
Comment (hallgrimurhg) [inline extensions.c:134]: Can we make blob a const here? Just to make it more explicit that it's a read-only input parameter.

**Review by hallgrimurhg — CHANGES_REQUESTED** [review]
Sorry I know this is meant as a refactoring, but noticed a couple of issues that might be good to address. Also a few nits. Thanks!

File: certificates.c [hunk 7]
```diff
@@ -269,17 +119,25 @@ hibacert_parse(struct hibacert *cert, struct sshkey *key) {
...
 				ret = HIBA_INTERNAL_ERROR;
 				goto err;
 			}
-			if ((ret = hibacert_load_extensions(cert, extension)) < 0) {
+			if ((ret = sshbuf_froms(extension, &data)) < 0) {
+				debug3("hibacert_parse: sshbuf_froms returned %d: %s", ret, ssh_err(ret));
+				ret = HIBA_INTERNAL_ERROR;
+				goto err;
+			}
+			if ((cert->nexts = hibaext_decode_all(&cert->exts, data)) < 0) {
```
[the remark this answers is not in the document]
Comment (blunderer) [reply certificates.c:135]: Done for the HIBA_INTERNAL_EROR, but there is no double free (it's `extension` versus `extensions`).

File: extensions.c [hunk 8]
```diff
@@ -40,6 +42,94 @@ struct hibaext {
...
+		stream.next_out = buf;
+		stream.avail_out = sizeof(buf);
+		ret = inflate(&stream, Z_SYNC_FLUSH);
+		if (ret != Z_OK && ret != Z_STREAM_END) {
+			break;
+		}
+		if (decomp == NULL) {
+			decomp = sshbuf_new();
+		}
+		sshbuf_put(decomp, buf, stream.total_out);
```
[the remark this answers is not in the document]
Comment (blunderer) [reply extensions.c:72]: Done

File: extensions.c [hunk 9]
```diff
@@ -40,6 +42,94 @@ struct hibaext {
 	struct pair pairs;
 };
 
+static struct sshbuf*
+hibaext_maybe_inflate(struct sshbuf *extensions) {
+	int ret = HIBA_OK;
```
[the remark this answers is not in the document]
Comment (blunderer) [reply extensions.c:47]: Done

File: extensions.c [hunk 10]
```diff
@@ -40,6 +42,94 @@ struct hibaext {
...
+
+			stream.next_out = buf;
+			stream.avail_out = sizeof(buf);
+			ret = deflate(&stream, Z_FINISH);
+			if (ret != Z_OK && ret != Z_STREAM_END) {
+				debug2("hibaext_maybe_deflate: deflate returned %d", ret);
+				ret = HIBA_INTERNAL_ERROR;
+				goto err;
+			}
+			sshbuf_put(temp, buf, stream.total_out);
```
[the remark this answers is not in the document]
Comment (blunderer) [reply extensions.c:113]: Done

File: extensions.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (blunderer) [reply extensions.c:72]: Done

File: extensions.c (hunk 3 above)
Comment (blunderer) [reply extensions.c:232]: Done

File: extensions.c (hunk 5 above)
Comment (blunderer) [reply extensions.c:293]: "MULT" is actually the value of the magic header. I added quotes around to make it clearer.

File: extensions.c (hunk 4 above)
Comment (blunderer) [reply extensions.c:277]: Good catch. I assumed the code would use the goto err, but it breaks instead in order to run the proper free calls.

Done

**Review by hallgrimurhg — APPROVED** [review]
Thanks, LGTM

**Pull request merged**